### PR TITLE
chore: add containerized deploy action

### PR DESCRIPTION
**Pull request by rokroskar**
continuation of #1751 

Note that the teardown workflow has been moved to the deploy & test workflow - it ran successfully [here](https://github.com/SwissDataScienceCenter/renku/runs/1687992079?check_suite_focus=true) for #1751.

/deploy

### CODE REVIEW

File: actions/deploy-renku/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/bin/sh
+
+RENKU_NAMESPACE=${RENKU_NAMESPACE:-$RENKU_RELEASE}
+RENKU_TMP_NAMESPACE=${RENKU_TMP_NAMESPACE:-${RENKU_NAMESPACE}-tmp}
```
Comment (lorenzo-cavazzi) [inline actions/deploy-renku/entrypoint.sh:4]: This is very convenient!
Should it be extended to the `teardown-renku` action as well?

**Review by lorenzo-cavazzi — CHANGES_REQUESTED** [review]
I really love this new ci action! Even if it's requiring some time, it will save much more once it's ready! :clock1: 

I added one minor comment inline.

I'm not sure the teardown action is working. I tried it here SwissDataScienceCenter/renku-ui#1196, the action succeeded but the deployment was not removed. I can still get the pods from `kubectl`

```
> kubectl -n ci-ui-1196 get pods

NAME                                            READY   STATUS                       RESTARTS   AGE
ci-ui-1196-core-6744c8db95-4vbtc                4/4     Running                      0          14m
ci-ui-1196-core-redis-master-0                  1/1     Running                      0          14m
ci-ui-1196-event-log-7bbddd4f6f-xfj27           0/1     CreateContainerConfigError   0          14m
ci-ui-1196-gateway-6ccc8b56c-dz2jg              1/1     Running                      0          14m
ci-ui-1196-gateway-auth-68d9c4c5c9-82bb6        1/1     Running                      3          14m
ci-ui-1196-gw-redis-master-0                    1/1     Running                      0          14m
ci-ui-1196-jena-6ddcf4bc9b-gfrs8                1/1     Running                      0          14m
ci-ui-1196-keycloak-0                           1/1     Running                      0          14m
ci-ui-1196-knowledge-graph-548ffb49b6-g2jvb     1/1     Running                      0          14m
ci-ui-1196-notebooks-6fcfccdf78-nlcnh           1/1     Running                      0          14m
ci-ui-1196-post-install-keycloak-hvfxs          0/1     Completed                    0          13m
ci-ui-1196-post-install-postgres-cpj2b          0/1     Completed                    0          14m
ci-ui-1196-postgresql-0                         2/2     Running                      1          14m
ci-ui-1196-token-repository-fcfc8fb79-kgcf8     0/1     CreateContainerConfigError   0          14m
ci-ui-1196-triples-generator-5b649c99b8-zj9rp   1/1     Running                      0          14m
ci-ui-1196-ui-5d77b5795d-4pmwt                  1/1     Running                      0          14m
ci-ui-1196-webhook-service-5c45996c6-c7f5f      1/1     Running                      0          14m
hub-78b8b57fdf-42ccv                            1/1     Running                      1          14m
proxy-567b7856b-vv6ps                           1/1     Running                      0          14m
```

Looking closer at the output (https://github.com/SwissDataScienceCenter/renku-ui/runs/1704726143?check_suite_focus=true), it seems that an error does occur, even if it doesn't report the action as failed

```
The connection to the server localhost:8080 was refused - did you specify the right host or port?
The connection to the server localhost:8080 was refused - did you specify the right host or port?
jq: error (at <stdin>:0): Cannot index string with string "application_name"
```

File: actions/deploy-renku/deploy_tmp_notebooks.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
Comment (ableuler) [inline actions/deploy-renku/deploy_tmp_notebooks.py:1]: Moving the file here broke the permissions, it should be an executable.

File: actions/deploy-renku/deploy_tmp_notebooks.py [hunk 3]
```diff
@@ -243,71 +334,23 @@ def main():
...
+    deploy_tmp_notebooks(
+        args.release_name,
+        args.renku_namespace,
+        renku_notebooks_version,
+        tmp_namespace,
+        cleanup=args.cleanup,
+        deploy=args.deploy,
+        extra_values=args.extra_values,
+        kube_context=args.kube_context,
+        local_path=None,
```
Comment (ableuler) [inline actions/deploy-renku/deploy_tmp_notebooks.py:351]:
```suggestion
        local_path=args.local_path,
```

File: actions/deploy-renku/deploy_tmp_notebooks.py [hunk 4]
```diff
@@ -187,16 +192,95 @@ def check_notebooks_chart_version():
...
+        f"{tmp_values_path}",
+        "--timeout",
+        "1800s",
+        "--kube-context",
+        f"{kube_context}",
+    ]
+
+    # check if the notebooks version is a local path
+    if local_path:
+        helm_command += ["renku-notebooks", str(local_path)]
```
Comment (ableuler) [inline actions/deploy-renku/deploy_tmp_notebooks.py:259]:
```suggestion
        helm_command += [str(local_path)]
```
I think when providing a local chart you don't give the name of the chart to that command (take `helm upgrade --install andreas-renku ./renku` for example).

**Review by ableuler — CHANGES_REQUESTED** [review]
Almost there - had to adapt 3 minor glitches to make this work outside of the context of a GH action for renku-notebooks.

File: actions/deploy-renku/deploy_tmp_notebooks.py (hunk 4 above)
Comment (rokroskar) [reply actions/deploy-renku/deploy_tmp_notebooks.py:259]: hmm I suppose you're right though I was pretty sure I tested this :)

File: actions/deploy-renku/entrypoint.sh (hunk 1 above)
Comment (rokroskar) [reply actions/deploy-renku/entrypoint.sh:4]: good point, thanks!

**Review by ableuler — APPROVED** [review]

**Comment by rokroskar** [comment]
Thanks @lorenzo-cavazzi and @ableuler!

@lorenzo-cavazzi and I discussed moving the conditional on the deploy step to the job, so that if `/deploy` is not present in the PR description the deploy and test steps are skipped altogether - this would prevent potentially confusing messages and environment definitions from appearing. I think this makes sense in general but not for the main renku repo, where we want to run acceptance tests for automatically merging PRs that update dependencies. An example of the setup I have in mind for other repos is [here](https://github.com/SwissDataScienceCenter/renku-python/pull/1781/files#diff-bc7672a2d6ff6960e810366c637bf2439aa87f10ab775dc25c0dfe51d60ab02dR13-R25).

**Review by lorenzo-cavazzi — APPROVED** [review]
:tada:  :rocket:  :exploding_head:

**Review by ableuler — APPROVED** [review]

**Comment by rokroskar** [comment]
I added the `automerge` label so that, in the off chance that the tests pass, this will get merged and stop me from making more commits

**Pull request merged**